gson/src/main/java/com/google/gson/stream/JsonWriter.java

```diff
@@ -638,15 +638,17 @@ public JsonWriter value(Number value) throws IOException {
 
     writeDeferredName();
     String string = value.toString();
-    if (string.equals("-Infinity") || string.equals("Infinity") || string.equals("NaN")) {
-      if (strictness != Strictness.LENIENT) {
-        throw new IllegalArgumentException("Numeric values must be finite, but was " + string);
-      }
-    } else {
-      Class<? extends Number> numberClass = value.getClass();
+    Class<? extends Number> numberClass = value.getClass();
+
+    if (!alwaysCreatesValidJsonNumber(numberClass)) {
       // Validate that string is valid before writing it directly to JSON output
-      if (!isTrustedNumberType(numberClass)
-          && !VALID_JSON_NUMBER_PATTERN.matcher(string).matches()) {
+      if (string.equals("-Infinity") || string.equals("Infinity") || string.equals("NaN")) {
+        if (strictness != Strictness.LENIENT) {
+          throw new IllegalArgumentException("Numeric values must be finite, but was " + string);
+        }
+      } else if (!(numberClass == Float.class
```

Comment:
I think I would be inclined to apply De Morgan here, similar to the way the condition was before:
```java
if (numberClass != Float.class
    && numberClass != Double.class
    && !VALID_JSON_NUMBER_PATTERN.matcher(string).matches())
```